Introduce radio manager+factory for RFM9x

nateinaction opened the pull request:
## Summary
This PR introduces radio factory and radio manager to handle calls to the RFM9x radio.
The factory handles all calls directly interacting with the adafruit library while the manager holds the instance of the radio and manage interactions between the radio and its callers. 

## How was this tested
- [x] Added new unit tests
- [x] Ran code on hardware (screenshots are helpful)
- [ ] Other (Please describe)
<img width="1552" alt="Screenshot 2025-02-15 at 12 01 26" src="https://github.com/user-attachments/assets/9f7b5dfd-81ec-47b1-8482-49c6ca28dd2d" />

nateinaction reviewed (commented):
I don't think I'm going to get the test coverage any higher. With the free version of SonarQube I cannot create my own "quality gate" with a lower test coverage required. We'll have to either:
1. live with these failures until we add more testing
2. remove code coverage reporting in SonarQube
3. stop using SonarQube

As much as I dislike seeing these failures, they're a good push. I even went back and added a few more tests to try and satisfy it... I got close.

Mikefly123 commented:
> I don't think I'm going to get the test coverage any higher. With the free version of SonarQube I cannot create my own "quality gate" with a lower test coverage required. We'll have to either:
> 
> 1. live with these failures until we add more testing
> 2. remove code coverage reporting in SonarQube
> 3. stop using SonarQube
> 
> As much as I dislike seeing these failures, they're a good push. I even went back and added a few more tests to try and satisfy it... I got close.

I'm pretty new to these DevOps tools, but would be perhaps consider self-hosting the SonarQube Community Edition to get around this? It's possible we could set up an environment in one of our labs to oversee this kind of stuff.

Mikefly123 reviewed (changes requested):
This PR looks great @nateinaction! I only have 1 requested change on that `low_datarate_optimize` thing that needs to be updated before merging. Most of my other comments are more just food for thought that we can wrench on with future PRs. 

Looking forward to seeing this get into the main codebase!

nateinaction commented on lib/pysquared/pysquared.py:
```diff
@@ -384,7 +330,6 @@ def orpheus_init_UART(hardware_key: str):
         #                                         #
         ######## Temporary Fix for RF_ENAB ########
 
-        self.init_radio(hardware_key="Radio1")
```
I agree that this is a sore spot. We are starting to shift our eyes to the REPL/ground station experience but we haven't spent much time on it. I have a few ideas on how to solve this.

1. Like you, I think we need some REPL tooling that's ready to go out of the box. No special file drag/drop needed, no long instantiation requirements, etc. `make install` (or install from a zip from github) and you're done!
2. I think we can start work on a dependency injection framework that can help automate some of this setup.

Lots of fun stuff to do in this space.

Mikefly123 replied:
I went ahead and started #182 to further track this stuff!

nateinaction commented on lib/pysquared/hardware/rfm9x/manager.py:
```diff
@@ -0,0 +1,123 @@
+        self._transmit_power = transmit_power
+        self._lora_spreading_factor = lora_spreading_factor
+
+        self._radio = self.radio
```
Can you tell me more about why we would want to have more than one radio instantiated? I suspect we will eventually want radios of different bands/capabilities that might have their own managers but I'm not sure I'm seeing why we would have multiple RFM9x radios on board.

We could extend this to hold multiple radios but I'd like to understand the usecase and when we might expect to see boards/satellites with this functionality?

On "clunky" feedback: I think we will get to the point where the functionality of the upstream radio library will be obscured by the radio manager so the call would be something closer to:
```
radio_manager.send(data)
```

Once we better define how we want to perform message chunking and other specialized features we can start to add this functionality to the manager.

Mikefly123 replied:
We discussed this a bit in the last software working group meeting, so I just wanted to put some notes here for documentation before resolving this conversation:

- The reason to want multiple radios instantiated is to have access to multiple bands. One of the interesting parts of this is that for both the 70cm (435Mhz) and 33cm (915Mhz) band we are able to use a common RFM9x library to drive them.
- The upcoming V5 flight controller dev boards will be the first set of boards to have two radios that can each be tuned to 70cm and 33cm respectively.
- In flight the reason to do this would potentially be to only use on band for transmitting and the other band only for receiving.
- I think for now we can kick this can down the road and determine after getting some more experience with new hardware whether it makes sense to expand this radio manager with multi-band capabilities or if you would create a dedicated manager for each radio instance.

nateinaction commented on lib/pysquared/hardware/exception.py:
```diff
@@ -0,0 +1,2 @@
+class HardwareInitializationError(Exception):
+    pass
```
This is the recommended method to create custom exception types in python.

nateinaction commented:
> I'm pretty new to these DevOps tools, but would be perhaps consider self-hosting the SonarQube Community Edition to get around this? It's possible we could set up an environment in one of our labs to oversee this kind of stuff.

100% can be done. I'm not sure if it's going to jump to the top of my priority list though. I think we can keep bring up test coverage pretty quickly so it's a temporary annoyance. We're already above 10% on `main` with more to come!

nateinaction commented on lib/pysquared/hardware/decorators.py:
```diff
@@ -0,0 +1,36 @@
+    """Decorator that retries hardware initialization with exponential backoff.
+
+    :param int max_attempts: Maximum number of attempts to try initialization (default is 3)
+    :param int initial_delay: Initial delay in seconds between attempts (default is 1.0)
```
Agreed, this is important to keep in mind. We really need to make sure the watch dog runs async from the rest of the control loop. I'm going to need to understand how to use `asyncio` a lot better than I do now...

Mikefly123 replied:
Adding a thought here that might explode the scope of work for all this dev by quite a bit.

Basically we might want to consider really leaning into the component model for software by making *all* of our discrete events on the satellites and `asyncio` task that is called at a particular time by a sequence manager. I think this is how F Prime currently operates, although with the benefit of true threads for real concurrency. 

We need to do more experimenting and understanding exactly what CircuitPython `asyncio` can and cannot do, but the general idea behind this would be to allow tasks to release the thread when they encounter an operation that might take a while or block the thread (like waiting for a radio receive or data to be collected) so things like the watchdog can still be pet on time.

nateinaction commented on lib/pysquared/hardware/digitalio.py:
```diff
@@ -0,0 +1,37 @@
+        digital_in_out.direction = direction
+        digital_in_out.value = initial_value
+        return digital_in_out
+    except Exception as e:
```
Hmmm you bring up a good point but I'm not sure if I would put a firmware check here. I wonder if we should perform a firmware check somewhere else? I wonder if we can do it on `make install` or on boot? Interesting idea...

In general though, heck yeah! If we can provide some kind of recovery behavior here that would be great but I agree that immutability is a concern.

Mikefly123 replied:
Hm I think it would be pretty cool if we could check for firmware on `make install`! I opened #181 to track that. I think I will say changing this error handling is out of scope for this pull request, so I will just resolve this and we can circle back at some later time.

nateinaction commented on lib/pysquared/hardware/rfm9x/factory.py:
```diff
@@ -0,0 +1,168 @@
+
+        radio.ack_delay = 0.2
+        radio.enable_crc = True
+        radio.max_output = True
```
Ok, we can pull these from config. 👍

nateinaction commented:
@Mikefly123 Thanks for your patience, this is ready for rereview. I addressed feedback on `low_datarate_optimize` and moving hardcoded values to config.

I have replied with questions on your other feedback to hopefully help me understand your mindset. Ultimately I think we are aligned that there are still painpoints and that we are introducing painpoint for satellite operators with extractions like these. I know we can make progress on resolving some of those but I'll need help where I lack understanding.

Mikefly123 commented:
> @Mikefly123 Thanks for your patience, this is ready for rereview. I addressed feedback on `low_datarate_optimize` and moving hardcoded values to config.
> 
> I have replied with questions on your other feedback to hopefully help me understand your mindset. Ultimately I think we are aligned that there are still painpoints and that we are introducing painpoint for satellite operators with extractions like these. I know we can make progress on resolving some of those but I'll need help where I lack understanding.

Hey @nateinaction! Yeah I saw the comments come in yesterday and haven't quite gotten around to replying to them all yet. Sorry to stall getting this PR through, but I can commit to getting back on the horse with this over the weekend after we get through our big design review for SCALES tomorrow!

Mikefly123 reviewed (approved):
Okay this all looks good to me! 

I am on travel right now, so I don't have the hardware to do and end to end test. If we can quickly verify that the new code behaves and transmits as expected let's merge this thing!

nateinaction commented:
Output looks normal: 
<img width="1552" alt="Screenshot 2025-03-05 at 22 10 57" src="https://github.com/user-attachments/assets/f3e9aafc-db40-4ae1-93d0-db1a86d67617" />
